### PR TITLE
Artur Wodarz - Solution

### PR DESCRIPTION
**Pull request by ArtWodarz**

### CODE REVIEW

File: test/galaxy.test.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        expect(result).toBeInstanceOf(Array);
+        expect(result.length).toBe(29);
+        result.forEach(el => expect(el).toBeInstanceOf(SpaceObject));
+    });
+    it("Should return proper objects", () => {
+        const result = galaxy.generateObjects();
+        let nofK = 0;
+        let nofSB = 0;
+        let nofS = 0;
+        result.forEach(el => {
```
Comment (firemark) [inline test/galaxy.test.js]: much cleaner would be to make a count function:

```js
function count(type) {
  return result.filter(o => o.type === type).length;
}
expect(count('ship')).toBe(config.FOO);

```

File: test/game.test.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            new SpaceObject(new Point(4, 4), new Point(3, 3), 'star')
+        ];
+        const us = ['?', '?', '?'];
+        const ks = [
+            [[2, 0, 0], [0, 1, 1], [0, 0, 0]],
+            [[0, 0, 0], [0, 0, 0], [0, 0, 0]],
+            [[0, 0, 0], [0, 0, 0], [0, 0, 0]]
+        ];
+        const expected = [
+            [ks[0][0], ks[0][1], ks[0][2], us, us, us, us, us],
```
Comment (firemark) [inline test/game.test.js]: nice :)

File: test/renderer.test.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+const SpaceShip = require("../src/core/spaceShip");
+const Point = require("../src/utils/point");
+
+const renderer = require("../src/utils/renderer");
+const eol = require("os").EOL;
+
+const player = new SpaceShip(new Point(3, 3), new Point(3, 3), 600);
+
+describe("Test renderEmptyQuadrant", () => {
+    it("Should display quadrant of given schema", () => {
```
Comment (firemark) [inline test/renderer.test.js]: amazing!

**Review by firemark — CHANGES_REQUESTED** [review]

File: src/core/galaxy.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    shuffleQuadrant(quadrant) {
+        let newRnds = MathSupport.randomUniques(
+            0,
+            this.config.SECTORS_IN_QUADRANT - 1,
+            this.getQuadrantObjects(quadrant).length
+        );
+        let currentRnd = 0;
+        this.SpaceObjects = this.SpaceObjects.map(so => {
+            if(so.quadrant.x == quadrant.x && so.quadrant.y == quadrant.y){
```
[the remark this answers is not in the document]
Comment (Bornegio) [reply src/core/galaxy.js]: Actually mathSupport module has hasSamePos() method, which compares two instances of point. Can be used with both quadrants and sectors. I just missed this line when I was refactoring code. Gonna change that. Thanks

File: src/core/game.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                    this.player.reducePower(dmg);
+                } else {
+                    console.log('Klingons has attacked, but missed.');
+                }
+            }
+        });
+    }
+
+    playerShoot(power){
+        if(power >= this.player.power){
```
Comment (imatw4r) [inline src/core/game.js]: Self destruction with `power > this.player.power` would be interesting option for a tie result.

File: src/core/game.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        let closestKlingons = this.galaxy.getQuadrantObjects(this.player.quadrant)
+            .filter(so => so.type == 'ship')
+            .sort((prev, cur) => {
+                let prevDist = mathSupport.cityBlockDistance(this.player.sector.x, this.player.sector.y, prev.sector.x, prev.sector.y);
+                let curDist = mathSupport.cityBlockDistance(this.player.sector.x, this.player.sector.y, cur.sector.x, cur.sector.y);
+                return prevDist > curDist;
+            });
+        if(closestKlingons.length <= 0){
+            console.log('There is no remaining klingons in your current quadrant');
```
Comment (imatw4r) [inline src/core/game.js]: I would add `return` statement here and get rid of `else` statement below. I like the approach with sticking to the rule: quick ASAP. Basically return from the function as fast as you can. If you can.

File: src/core/game.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        } else {
+            this.player.reducePower(power);
+            let distanceToPlayer = mathSupport.cityBlockDistance(
+                this.player.sector.x,
+                this.player.sector.y,
+                closestKlingons[0].sector.x,
+                closestKlingons[0].sector.y
+            );
+            let hasHit = Math.random() < (5 / (distanceToPlayer + 4));
+            if (hasHit){
```
Comment (imatw4r) [inline src/core/game.js]: Same here. You could do
```
if (!hasHit) {
    console.log("You missed!")
    return
}

closestKlingons[0].reducePower(power);
<rest of the code>
```
It definitely decreases the nesting levels and removes some IFs

File: src/core/game.js [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        }
+    }
+
+    detectPlayerCollisions() {
+        this.galaxy.getQuadrantObjects(this.player.quadrant).forEach(so => {
+            if (mathSupport.hasSamePos(so.sector, this.player.sector)) {
+                this.player.isDestroyed = true;
+                return;
+            } else if (so.type == "starbase") {
+                this.galaxy.getSurrouningSectors(so.sector).forEach(p => {
```
Comment (imatw4r) [inline src/core/game.js]: Typo `getSurrouningSectors` => `getSurroundingSectors`

File: src/index.js [hunk 9]
```diff
@@ -1,44 +1,71 @@
...
-let mode = COMMANDS_MODE; // the current mode
+let mode = COMMANDS_MODE;
+let game = new Game(config);
 
-while (true)
-{
-    if (mode === COMMANDS_MODE) { // the normal commands mode
-        
+while (!game.isOver) {
+    game.checkIfOver();
```
Comment (imatw4r) [inline src/index.js]: Is it necessary to keep it here? Shouldn't it be at the end of the loop? I would expect the end game check after each move not before it.

**Review by imatw4r — COMMENTED** [review]
Your code looks pretty clean, I really like it! 
Even the variable names are pretty good!
Keep up the good work!

**Pull request closed without merging**